sdk/core/core/inc/az_action.h

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_ACTION_H
+#define AZ_ACTION_H
+
+#include <az_contract.h>
+
+#include <_az_cfg_prefix.h>
+
+#define AZ_CAT(A, B) A##B
+
+#define AZ_ACTION_ARG(NAME) AZ_CAT(NAME, _arg)
+
+/**
+ * Defines an action @NAME type which accepts an argument of type @ARG.
+ */
+#define AZ_ACTION_TYPE(NAME, ARG) \
+  typedef ARG AZ_ACTION_ARG(NAME); \
```

Comment:
I really don't like putting this stuff in a macro, especially one that defines structs.

Reply:
Me neither, but I don't know a better way to do it without buggy unsafe `void *` casting.